"Internal errror: shutdownContext is null" During complex tests

### Describe the bug

I have a codebase with a series of tests, spanning different test profiles and thus restarts Quarkus several times.

Upgrading from Quarkus `3.21.4` to `3.29.0` results in sporadic but frequent and consistent errors when these restarts occur; 

```
java.lang.RuntimeException: Failed to start quarkus
java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:666)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:761)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:276)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:276)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:276)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:351)
	at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:232)
	at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:640)
	at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:685)
	... 1 more
Caused by: java.lang.RuntimeException: Internal errror: shutdownContext is null. This probably happened because Quarkus failed to start properly in an earlier step, or because tests were run on a Quarkus instance that had already been shut down.
	at io.quarkus.runtime.configuration.ConfigRecorder.releaseConfig(ConfigRecorder.java:113)
	at io.quarkus.runner.recorded.ConfigGenerationBuildStep$releaseConfigOnShutdown561040398.deploy_0(Unknown Source)
	at io.quarkus.runner.recorded.ConfigGenerationBuildStep$releaseConfigOnShutdown561040398.deploy(Unknown Source)
	... 8 more


Failed to start quarkus
java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:351)
	at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:232)
	at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:640)
	at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:685)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.lang.RuntimeException: Internal errror: shutdownContext is null. This probably happened because Quarkus failed to start properly in an earlier step, or because tests were run on a Quarkus instance that had already been shut down.
	at io.quarkus.runtime.configuration.ConfigRecorder.releaseConfig(ConfigRecorder.java:113)
	at io.quarkus.runner.recorded.ConfigGenerationBuildStep$releaseConfigOnShutdown561040398.deploy_0(Unknown Source)
	at io.quarkus.runner.recorded.ConfigGenerationBuildStep$releaseConfigOnShutdown561040398.deploy(Unknown Source)
	... 8 more

```

### Expected behavior

`java.lang.RuntimeException: Internal errror: shutdownContext is null.` shouldn't occur during tests

### Actual behavior

`java.lang.RuntimeException: Internal errror: shutdownContext is null.` occurs during tests

### How to Reproduce?

Reproducer here: https://github.com/Epic-Breakfast-Productions/OpenQuarterMaster/tree/dev/core-api-q-upgrade

The project in question is in the `software/core/oqm-core-api` directory. On the specified branch, I have already done the Quarkus version upgrade and some cleanup of test resources, attempting to simplify and narrow possibilities.

`./gradlew test`

Notes:
 - Note the `core-api-q-upgrade` branch
 - There is a known issue with Lombok that sometimes blocks compilation. This clears if you retry building (maybe a few times). Might require a clean if persistent (sorry, I have a bug out with the Lombok group, but it's not easily reproducible)

### Output of `uname -a` or `ver`

Linux oqm-dev 6.14.0-33-generic #33~24.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Sep 19 17:02:30 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

21

### Quarkus version or git rev

`3.21.4` upgrading to `3.28.4`

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 9.1.0

### Additional information

Please let me know what else I could provide.

--- a/core/deployment/src/main/java/io/quarkus/runner/bootstrap/StartupActionImpl.java
+++ b/core/deployment/src/main/java/io/quarkus/runner/bootstrap/StartupActionImpl.java
@@ -324,6 +324,13 @@ public class StartupActionImpl implements StartupAction {
      * Runs the application, and returns a handle that can be used to shut it down.
      */
     public RunningQuarkusApplication run(String... args) throws Exception {
+
+        if (runtimeClassLoader.isClosed()) {
+            throw new RuntimeException(
+                    "Internal error: An attempt was made to start an application which had already been started and closed. The affected ClassLoader is "
+                            + runtimeClassLoader);
+
+        }
         // Start dev services that weren't started in the augmentation phase
         ensureDevServicesStarted();
         InitialConfigurator.DELAYED_HANDLER.buildTimeComplete();

--- a/core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigRecorder.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigRecorder.java
@@ -113,10 +113,6 @@ public class ConfigRecorder {
         // it actually does not because it operates on a different instance
         // of QuarkusConfigFactory from a different classloader.
 
-        if (shutdownContext == null) {
-            throw new RuntimeException(
-                    "Internal error: shutdownContext is null. This probably happened because Quarkus failed to start properly in an earlier step, or because tests were run on a Quarkus instance that had already been shut down.");
-        }
         shutdownContext.addLastShutdownTask(QuarkusConfigFactory::releaseTCCLConfig);
     }
 }

--- a/test-framework/junit/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
+++ b/test-framework/junit/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
@@ -625,9 +625,10 @@ public class QuarkusTestExtension extends AbstractJvmQuarkusTestExtension
             ClearCache.clearCaches();
         }
 
-        // TODO if classes are misordered, say because someone overrode the ordering, and there are profiles or resources,
-        // we could try to start and application which has already been started, and fail with a mysterious error about
-        // null shutdown contexts; we should try and detect that case, and give a friendlier error message
+        if (cl.isClosed()) {
+            throw new IllegalStateException(
+                    "Internal error. Attempting to run tests using an application which has already been closed. Is a non-default test order being used?");
+        }
 
         // We want to start if the profile changed (or there are new test resources),
         // or if we don't have an app and that's not because the previous attempt to start it failed

--- a/test-framework/junit/src/main/java/io/quarkus/test/junit/classloading/FacadeClassLoader.java
+++ b/test-framework/junit/src/main/java/io/quarkus/test/junit/classloading/FacadeClassLoader.java
@@ -641,6 +641,14 @@ public final class FacadeClassLoader extends ClassLoader implements Closeable {
         return isServiceLoaderMechanism;
     }
 
+    /**
+     * Returns true if, after test discovery, multiple classloaders have been created. This would mean multiple Quarkus
+     * applications will be started to run the tests.
+     */
+    public boolean hasMultipleClassLoaders() {
+        return runtimeClassLoaders != null && runtimeClassLoaders.size() > 1;
+    }
+
     @Override
     public String getName() {
         return NAME;
@@ -659,5 +667,4 @@ public final class FacadeClassLoader extends ClassLoader implements Closeable {
         runtimeClassLoaders.clear();
 
     }
-
 }

--- a/test-framework/junit/src/main/java/io/quarkus/test/junit/launcher/CustomLauncherInterceptor.java
+++ b/test-framework/junit/src/main/java/io/quarkus/test/junit/launcher/CustomLauncherInterceptor.java
@@ -1,5 +1,8 @@
 package io.quarkus.test.junit.launcher;
 
+import java.util.Optional;
+
+import org.junit.jupiter.api.ClassOrderer;
 import org.junit.platform.launcher.LauncherDiscoveryListener;
 import org.junit.platform.launcher.LauncherDiscoveryRequest;
 import org.junit.platform.launcher.LauncherSession;
@@ -7,10 +10,13 @@ import org.junit.platform.launcher.LauncherSessionListener;
 import org.junit.platform.launcher.TestExecutionListener;
 import org.junit.platform.launcher.TestPlan;
 
+import io.quarkus.test.config.QuarkusClassOrderer;
 import io.quarkus.test.junit.classloading.FacadeClassLoader;
 
 public class CustomLauncherInterceptor
         implements LauncherDiscoveryListener, LauncherSessionListener, TestExecutionListener {
+
+    private static final Class<? extends ClassOrderer> DESIRED_CLASS_ORDERER = QuarkusClassOrderer.class;
 
     private static FacadeClassLoader facadeLoader = null;
     // Also use a static variable to store a 'first' starting state that we can reset to
@@ -53,7 +59,6 @@ public class CustomLauncherInterceptor
             origCl = Thread.currentThread()
                     .getContextClassLoader();
         }
-        // We might not be in the same classloader as the Facade ClassLoader, so use a name comparison instead of an instanceof
         initializeFacadeClassLoader();
         adjustContextClassLoader();
 
@@ -125,7 +130,40 @@ public class CustomLauncherInterceptor
                 Thread.currentThread().setContextClassLoader(origCl);
 
             }
+
+            Optional<String> orderer = request.getConfigurationParameters().get("junit.jupiter.testclass.order.default");
+
+            if (orderer.isEmpty() || !orderer.get().equals(DESIRED_CLASS_ORDERER.getName())) {
+                if (facadeLoader.hasMultipleClassLoaders()) {
+                    String message = getFailureMessageForJUnitMisconfiguration(orderer);
+                    throw new IllegalStateException(message);
+                }
+            }
         }
+    }
+
+    private static String getFailureMessageForJUnitMisconfiguration(Optional<String> orderer) {
+        String generalExplanation = """
+                Critical failure. Quarkus tests would fail with corrupted application errors.
+                The reason is that they would not run in the right order, because the Quarkus JUnit configuration has been overridden.
+                When there are multiple test profiles or resources, Quarkus uses a JUnit ClassOrderer to sort tests so that test with the same profile run one after each other.
+                Running tests with a different sorting risks tests running on an application that has been cleaned up.
+                """;
+
+        String message;
+        if (orderer.isPresent()) {
+            message = String.format(
+                    "%sTo set a test order while preserving the Quarkus required sorting, please set use the Quarkus configuration to set junit.quarkus.orderer.secondary-orderer=%s, and remove the junit-platform.properties (if any) from the classpath.",
+                    generalExplanation, orderer.get());
+        } else {
+            message = String.format(
+                    """
+                            %sIt looks like there is a junit-platform.properties configuration file on the project classpath.
+                            The JUnit framework will only read the first properties it finds, which prevents Quarkus from setting the class orderer it needs.
+                            Please either add junit.jupiter.testclass.order.default=%s to your project's junit-platform.properties file, or remove the file and use system properties to configure JUnit.""",
+                    generalExplanation, DESIRED_CLASS_ORDERER.getName());
+        }
+        return message;
     }
 
     @Override

